### ISSUE TITLE
Vanilla PG log prob is no the original log prob?

### ISSUE THREAD

**Issue by drozzy** [thu-ml/tianshou#397]
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

I've been looking at the basic reinfoce/pg algorithm, and I don't understand why the implementation here seems to get a new set of log_prob values (by calling forward again) - instead of re-using the log_prob values that were saved into the buffer:
https://github.com/thu-ml/tianshou/blob/ebaca6f8da91e18e0192184c24f5d13e3a5d0092/tianshou/policy/modelfree/pg.py#L126

For me these two produce different outcomes, since a different action may be sampled on the second time a `forward` is called.

If my question is not clear please tell me how I can make it clearer. Thank you.

**Comment by Trinkle23897** [comment]
Hmm...Actually that's adapted from PPO.

**Comment by drozzy** [comment]
Hm.. ok I see what you are doing here.
You are storing data in the replay buffer, but without the gradient. So that is why you need to feed it through the NN again.
But one thing I don't understand is - the reward that was received from the environment was for action `b.act`, while during updating you sample an action `a` which may be a different action:

https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L128

So you will end up boosting the wrong action. Maybe I misunderstood something.

**Comment by ChenDRAG** [comment]
@drozzy  Actually no. L128 only converts data type of b.act(which is sampled from environment) to data type of result.act(which is the newly sampled action). The value of result.act is never actually used. See Line
https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L130

**Comment by ChenDRAG** [comment]
> You are storing data in the replay buffer, but without the gradient. So that is why you need to feed it through the NN again.

This is correct. Because we have to store all information in `Buffer`, which is of data type `numpy.ndarray`. We currently haven't figured out an efficient way to convert torch based gradient information to numpy array.
However, to recompute the action doesn't really take much time. You can run an experiment to test that if you want.

**Comment by drozzy** [comment]
Oooooh!
I get it. So you store `dist` just so that you can recover `log_prob` with a gradient afterwards.
This would be equivalent to just storing `logits` and then constructing a distribution from that and evaluating a specific action using it afterwards.
Just tried this and it works. Awesome!

**Comment by drozzy** [comment]
I tried to simplify the PG implementation to be as close as possible to my simple pytorch implementation and this is the result:
https://github.com/drozzy/reinforce_tianshou/blob/main/reinforce_tianshou.py

I found the tianshou PG implementation a bit too advanced to learn from for me :-) Hopefully this will help someone else in the future as well.
